src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -4,67 +4,186 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
-using Microsoft.Extensions.ObjectPool;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.AspNetCore.Razor.PooledObjects;
 
 /// <summary>
 ///  Wraps a pooled <see cref="ImmutableArray{T}.Builder"/> but doesn't allocate it until
 ///  it's needed. Note: Dispose this to ensure that the pooled array builder is returned
 ///  to the pool.
+///
+///  There is significant effort to avoid retrieving the <see cref="ImmutableArray{T}.Builder"/>.
+///  For very small arrays of length 4 or less, the elements will be stored on the stack. If the array
+///  grows larger than 4 elements, a builder will be employed.
 /// </summary>
-internal ref partial struct PooledArrayBuilder<T>
+internal partial struct PooledArrayBuilder<T> : IDisposable
```

Comment:
💡 You'll need to add the equivalent of this:
https://github.com/dotnet/roslyn-analyzers/blob/main/src/Roslyn.Diagnostics.Analyzers/Core/TemporaryArrayAsRefAnalyzer.cs

Reply:
Interesting! I do wish this analyzer was written a bit more generally so that it could be re-used, but it doesn't seem like it'd be too much effort to add to Razor.

Reply:
Added this to Razor -- thanks!